File: wear/src/main/java/com/asamm/locus/addon/wear/gui/MainSettingsActivity.java
/*
 * Created by milan on 10.08.2018.
 * This code is part of Locus project from Asamm Software, s. r. o.
 * Copyright (C) 2018
 */
package com.asamm.locus.addon.wear.gui;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.Switch;
import android.widget.TextView;
import android.widget.Toast;

import com.asamm.locus.addon.wear.R;
import com.asamm.locus.addon.wear.application.PreferencesEx;
import com.asamm.locus.addon.wear.application.FeatureConfigEnum;
import com.asamm.locus.addon.wear.application.TrackRecordingService;
import com.asamm.locus.addon.wear.common.communication.DataPath;
import com.asamm.locus.addon.wear.common.communication.containers.DataPayload;
import com.asamm.locus.addon.wear.common.communication.containers.commands.EmptyCommand;
import com.asamm.locus.addon.wear.features.trackRecord.recording.sensors.RecordingSensorManager;

public class MainSettingsActivity extends LocusWearActivity {

    private Switch switchHrm;

    @Override
    protected DataPayload<EmptyCommand> getInitialCommandType() {
        return null;
    }

    @Override
    protected DataPath getInitialCommandResponseType() {
        return null;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main_settings);
        TextView header = findViewById(R.id.text_view_screen_header);
        if (header != null) {
            header.setText(getText(R.string.settings));
        }
        switchHrm = findViewById(R.id.switch_hrm);

        updateSwitchState();
    }

    private void updateSwitchState() {
        FeatureConfigEnum hrmConfig = PreferencesEx.getHrmFeatureConfig();
        switchHrm.setChecked(hrmConfig == FeatureConfigEnum.ENABLED);
        switchHrm.setTextColor(hrmConfig != FeatureConfigEnum.NOT_AVAILABLE ? getColor(R.color.base_dark_primary) : getColor(R.color.base_light_disabled));
    }

    private boolean isHrmEnabled() {
        FeatureConfigEnum hrmConfig = PreferencesEx.getHrmFeatureConfig();
        return hrmConfig == FeatureConfigEnum.ENABLED;
    }

    /**
     * @return false for this activity, no handshaking required
     */
    @Override
    protected boolean isMakeHandshakeOnStart() {
        return false;
    }

    @Override
    public boolean isChildLocusWearActivity() {
        return true; // runs on top of TrackRecording activity, not independetly
    }

    @Override
    public void registerHwKeyActions(LocusWearActivityHwKeyDelegate delegate) {

    }

    public void onHrmClicked(View view) {
        if (isHrmEnabled()) {
            PreferencesEx.persistHrmFeatureConfig(FeatureConfigEnum.DISABLED);
            if (TrackRecordingService.isRunning()) {
                Intent intent = new Intent(this, TrackRecordingService.class);
                intent.setAction(TrackRecordingService.ACTION_STOP_FOREGROUND_SERVICE);
                startService(intent);
            }
        } else {
            if (RecordingSensorManager.checkAndRequestBodySensorPermission(this)) {
                enableHrm();
            }
        }
        updateSwitchState();
    }

    private void enableHrm() {
        runOnUiThread(() -> {
            FeatureConfigEnum hrmConfig = PreferencesEx.getHrmFeatureConfig();
            if (hrmConfig == FeatureConfigEnum.NO_PERMISSION) {
                if (RecordingSensorManager.checkBodySensorPermission(this)) {
                    PreferencesEx.persistHrmFeatureConfig(FeatureConfigEnum.NOT_AVAILABLE);
                } else {
                    Toast.makeText(this, getString(R.string.err_no_hrm_permission), Toast.LENGTH_LONG).show();
                }
            }
            if (hrmConfig == FeatureConfigEnum.NOT_AVAILABLE) {
                FeatureConfigEnum tmpHrmConfig = RecordingSensorManager.recheckSensorAvailability(this);
                // HRM could have been enabled as side effect of calling recheckSensorAvailability()
                if (tmpHrmConfig == FeatureConfigEnum.NOT_AVAILABLE) {
                    // sensor still not available even after recheck
                    Toast.makeText(this, getString(R.string.err_hrm_sensor_not_available), Toast.LENGTH_LONG).show();
                }
            }
            if (hrmConfig == FeatureConfigEnum.DISABLED) {
                // sensor should be accesible and is only disabled, simply enable it
                PreferencesEx.persistHrmFeatureConfig(FeatureConfigEnum.ENABLED);
            }
        });
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, String permissions[], int[] grantResults) {
        RecordingSensorManager.handlePermissionResult(this, requestCode, permissions, grantResults);
        enableHrm();
        updateSwitchState();
    }
}

File: wear/src/main/java/com/asamm/locus/addon/wear/features/settings/MainSettingsActivity.kt
/*
 * Created by milan on 10.08.2018.
 * This code is part of Locus project from Asamm Software, s. r. o.
 * Copyright (C) 2018
 */
package com.asamm.locus.addon.wear.features.settings

import android.content.Intent
import android.os.Bundle
import android.widget.Toast
import androidx.activity.compose.setContent
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.livedata.observeAsState
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.wear.compose.material.Icon
import androidx.wear.compose.material.MaterialTheme
import androidx.wear.compose.material.ToggleChipDefaults
import com.asamm.locus.addon.wear.R
import com.asamm.locus.addon.wear.common.communication.DataPath
import com.asamm.locus.addon.wear.common.communication.containers.DataPayload
import com.asamm.locus.addon.wear.common.communication.containers.commands.EmptyCommand
import com.asamm.locus.addon.wear.features.trackRecord.TrackRecordingService
import com.asamm.locus.addon.wear.features.trackRecord.recording.sensors.BodySensorsManager
import com.asamm.locus.addon.wear.gui.LocusWearActivity
import com.asamm.locus.addon.wear.gui.compose.*
import com.asamm.locus.addon.wear.gui.compose.theme.AppTheme
import com.asamm.locus.addon.wear.utils.FeatureConfigEnum
import com.asamm.logger.Logger

/**
 * Main app settings activity.
 */
class MainSettingsActivity : LocusWearActivity() {

    // runs from the menu, not independently
    override val isChildLocusWearActivity: Boolean
        get() = true

    override val initialCommandType: DataPayload<EmptyCommand>?
        get() = null

    override val initialCommandResponseType: DataPath?
        get() = null

    override val isMakeHandshakeOnStart: Boolean
        get() = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContent {
            MainContent()
        }
    }

    @Composable
    fun MainContent() {
        AppTheme {
            Column {
                Header()
                Content()
            }
        }
    }

    @Composable
    private fun Header() {
        Box(
            modifier = Modifier
                .fillMaxWidth()
                .height(48.dp)
                .background(MaterialTheme.colors.primary)
                .padding(bottom = 4.dp),
            contentAlignment = Alignment.BottomCenter
        ) {
            LText(
                text = stringResource(R.string.settings),
                color = MaterialTheme.colors.onPrimary,
                style = MaterialTheme.typography.title2
            )
        }
    }

    @Composable
    private fun ColumnScope.Content() {
        val scrollState = rememberScrollState()
        val hrmState by PreferencesEx.hrmFeatureConfigStateLd.observeAsState()

        Column(
            modifier = Modifier
                .fillMaxWidth()
                .weight(1.0f, fill = true)
                .verticalScroll(scrollState),
            horizontalAlignment = Alignment.CenterHorizontally
        ) {
            // version info
            val versionName = packageManager.getPackageInfo(packageName, 0).versionName
            LText(
                text = getString(R.string.version_name_X, versionName),
                modifier = Modifier
                    .padding(all = 8.dp)
            )

            // HRM settings
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(start = 16.dp, end = 16.dp)
                    .clickable {
                        onHrmClicked()
                    },
                verticalAlignment = Alignment.CenterVertically
            ) {
                LText(
                    text = stringResource(id = R.string.settings_hrm),
                    modifier = Modifier
                        .weight(1.0f)
                )
                val hrmEnabled = hrmState === FeatureConfigEnum.ENABLED
                Icon(
                    imageVector = ToggleChipDefaults.switchIcon(checked = hrmEnabled),
                    contentDescription = "",
                    modifier = Modifier.padding(start = 8.dp),
                    tint = if (hrmEnabled) MaterialTheme.colors.primary else Color.Gray
                )
            }
            if (PreferencesEx.isDebug) {
                LText(text = "State: $hrmState")
            }
        }
    }

    private fun onHrmClicked() {
        when (PreferencesEx.hrmFeatureConfigState) {
            FeatureConfigEnum.ENABLED -> {
                PreferencesEx.hrmFeatureConfigState = FeatureConfigEnum.DISABLED
                if (TrackRecordingService.isRunning()) {
                    val intent = Intent(this, TrackRecordingService::class.java)
                    intent.action = TrackRecordingService.ACTION_STOP_FOREGROUND_SERVICE
                    startService(intent)
                }
            }
            else -> {
                if (BodySensorsManager.checkAndRequestBodySensorsPermission(this)) {
                    // enable because we have a valid permission
                    enableHrm()
                }
            }
        }
    }

    private fun enableHrm() {
        runOnUiThread {
            val hrmState = PreferencesEx.hrmFeatureConfigState
            Logger.d(TAG, "enableHrm(), state: $hrmState")

            // check permission
            if (hrmState === FeatureConfigEnum.NO_PERMISSION) {
                if (BodySensorsManager.checkBodySensorsPermission(this)) {
                    // change just to "not available" to process in next step
                    PreferencesEx.hrmFeatureConfigState = FeatureConfigEnum.NOT_AVAILABLE
                } else {
                    Toast.makeText(
                        this,
                        getString(R.string.err_no_hrm_permission),
                        Toast.LENGTH_LONG
                    ).show()
                }
            }

            // check availability
            if (hrmState === FeatureConfigEnum.NOT_AVAILABLE) {
                val tmpHrmConfig = BodySensorsManager.recheckSensorAvailability(this)
                // HRM could have been enabled as side effect of calling recheckSensorAvailability()
                if (tmpHrmConfig === FeatureConfigEnum.NOT_AVAILABLE) {
                    // sensor still not available even after recheck
                    Toast.makeText(
                        this,
                        getString(R.string.err_hrm_sensor_not_available),
                        Toast.LENGTH_LONG
                    ).show()
                }
            }
            if (hrmState === FeatureConfigEnum.DISABLED) {
                // sensor should be accessible and is only disabled, simply enable it
                PreferencesEx.hrmFeatureConfigState = FeatureConfigEnum.ENABLED
            }
        }
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        BodySensorsManager.handlePermissionResult(this, requestCode)
        enableHrm()
    }

    @Preview(
        widthDp = WEAR_PREVIEW_DEVICE_WIDTH_DP,
        heightDp = WEAR_PREVIEW_DEVICE_HEIGHT_DP,
        apiLevel = WEAR_PREVIEW_API_LEVEL,
        uiMode = WEAR_PREVIEW_UI_MODE,
        backgroundColor = WEAR_PREVIEW_BACKGROUND_COLOR_BLACK,
        showBackground = WEAR_PREVIEW_SHOW_BACKGROUND
    )
    @Composable
    fun WearAppPreview() {
        MainContent()
    }

    companion object {

        // tag for logger
        private const val TAG = "MainSettingsActivity"
    }
}